[ICE]: pretty: index out of bounds: the len is 1 but the index is 1

more compact version of #153853 with different melting point
<!--
[31mICE[0m: Rustc ./b.rs '' 'thread 'rustc' ($TID) panicked at compiler/rustc_middle/src/ty/print/pretty.rs:3316:50: 'index out of bounds: the len is 1 but the index is 1'', 'thread 'rustc' ($TID) panicked at compiler/rustc_middle/src/ty/print/pretty.rs:3316:50: 'index out of bounds: the len is 1 but the index is 1''
File: /tmp/im/2/b.rs
-->

auto-reduced (treereduce-rust):
````rust
#[rustc_paren_sugar]
trait Tr {
    fn method();
}
fn main() {
    <u8 as Tr>::method()
}

````

original:
````rust
#![feature(unboxed_closures)]
#[rustc_paren_sugar]
trait Tr {
    fn method();
}
fn main() {
    <u8 as Tr>::method()
}

````

Version information
````
rustc 1.96.0-nightly (a0e206b56 2026-03-13)
binary: rustc
commit-hash: a0e206b5665d7c075cdc0ca5f3637a697e8ff55a
commit-date: 2026-03-13
host: x86_64-unknown-linux-gnu
release: 1.96.0-nightly
LLVM version: 22.1.0

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/a0e206b5665d7c075cdc0ca5f3637a697e8ff55a/compiler/rustc_middle/src/ty/print/pretty.rs#L3310-L3322

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0658]: unboxed_closures are still evolving
 --> /tmp/icemaker_global_tempdir.umiFuUeWNIrC/rustc_testrunner_tmpdir_reporting.ClOL7t4Bl8Hf/mvce.rs:1:1
  |
1 | #[rustc_paren_sugar]
  | ^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #29625 <https://github.com/rust-lang/rust/issues/29625> for more information
  = help: add `#![feature(unboxed_closures)]` to the crate attributes to enable
  = note: this compiler was built on 2026-03-13; consider upgrading it if it is out of date

error: the `#[rustc_paren_sugar]` attribute is a temporary means of controlling which traits can use parenthetical notation
 --> /tmp/icemaker_global_tempdir.umiFuUeWNIrC/rustc_testrunner_tmpdir_reporting.ClOL7t4Bl8Hf/mvce.rs:2:1
  |
2 | / trait Tr {
3 | |     fn method();
4 | | }
  | |_^
  |
  = help: add `#![feature(unboxed_closures)]` to the crate attributes to use it


thread 'rustc' (93460) panicked at /rustc-dev/a0e206b5665d7c075cdc0ca5f3637a697e8ff55a/compiler/rustc_middle/src/ty/print/pretty.rs:3316:50:
index out of bounds: the len is 1 but the index is 1
stack backtrace:
   0:     0x7fe74cd3012b - <<std[f332ca7ac2aab280]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[caaf70a19b46f696]::fmt::Display>::fmt
   1:     0x7fe74d41ccc8 - core[caaf70a19b46f696]::fmt::write
   2:     0x7fe74cd47136 - <std[f332ca7ac2aab280]::sys::stdio::unix::Stderr as std[f332ca7ac2aab280]::io::Write>::write_fmt
   3:     0x7fe74cd06218 - std[f332ca7ac2aab280]::panicking::default_hook::{closure#0}
   4:     0x7fe74cd235a3 - std[f332ca7ac2aab280]::panicking::default_hook
   5:     0x7fe74bd1f9bc - std[f332ca7ac2aab280]::panicking::update_hook::<alloc[4f779a1cfcd17fd2]::boxed::Box<rustc_driver_impl[12052c784fa38642]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7fe74cd23882 - std[f332ca7ac2aab280]::panicking::panic_with_hook
   7:     0x7fe74cd062d8 - std[f332ca7ac2aab280]::panicking::panic_handler::{closure#0}
   8:     0x7fe74ccfa829 - std[f332ca7ac2aab280]::sys::backtrace::__rust_end_short_backtrace::<std[f332ca7ac2aab280]::panicking::panic_handler::{closure#0}, !>
   9:     0x7fe74cd07d3d - __rustc[c06758f224b6ffb]::rust_begin_unwind
  10:     0x7fe7498f13bc - core[caaf70a19b46f696]::panicking::panic_fmt
  11:     0x7fe749ca5702 - core[caaf70a19b46f696]::panicking::panic_bounds_check
  12:     0x7fe74c388d64 - <rustc_middle[1620aa76878d0fde]::ty::print::pretty::TraitRefPrintSugared as rustc_middle[1620aa76878d0fde]::ty::print::Print<rustc_middle[1620aa76878d0fde]::ty::print::pretty::FmtPrinter>>::print
  13:     0x7fe74c38aa4a - <rustc_middle[1620aa76878d0fde]::ty::print::pretty::TraitRefPrintSugared as core[caaf70a19b46f696]::fmt::Display>::fmt
  14:     0x7fe74caffb39 - <rustc_trait_selection[9f54bb3d479049cd]::error_reporting::TypeErrCtxt>::on_unimplemented_components
  15:     0x7fe74cafed5d - <rustc_trait_selection[9f54bb3d479049cd]::error_reporting::TypeErrCtxt>::on_unimplemented_note
  16:     0x7fe74cb046e7 - <rustc_trait_selection[9f54bb3d479049cd]::error_reporting::TypeErrCtxt>::report_selection_error
  17:     0x7fe74b60b3a6 - <rustc_trait_selection[9f54bb3d479049cd]::error_reporting::TypeErrCtxt>::report_fulfillment_errors
  18:     0x7fe74d7523d0 - <rustc_hir_typeck[e7143c23707bf68d]::fn_ctxt::FnCtxt>::check_expr_call
  19:     0x7fe74d7323b1 - <rustc_hir_typeck[e7143c23707bf68d]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  20:     0x7fe74d72c8e8 - <rustc_hir_typeck[e7143c23707bf68d]::fn_ctxt::FnCtxt>::check_expr_block
  21:     0x7fe74d73238c - <rustc_hir_typeck[e7143c23707bf68d]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  22:     0x7fe74d7028b4 - rustc_hir_typeck[e7143c23707bf68d]::check::check_fn
  23:     0x7fe74e69743c - rustc_hir_typeck[e7143c23707bf68d]::typeck_with_inspect::{closure#0}
  24:     0x7fe74e695eb0 - rustc_query_impl[c3886bc306851868]::query_impl::typeck::invoke_provider_fn::__rust_begin_short_backtrace
  25:     0x7fe74d505066 - rustc_query_impl[c3886bc306851868]::execution::try_execute_query::<rustc_data_structures[672205533455fe64]::vec_cache::VecCache<rustc_span[5c41c3a85b90e419]::def_id::LocalDefId, rustc_middle[1620aa76878d0fde]::query::erase::ErasedData<[u8; 8usize]>, rustc_middle[1620aa76878d0fde]::dep_graph::graph::DepNodeIndex>, false>
  26:     0x7fe74d504a03 - rustc_query_impl[c3886bc306851868]::query_impl::typeck::execute_query_non_incr::__rust_end_short_backtrace
  27:     0x7fe74de84f3e - <rustc_middle[1620aa76878d0fde]::ty::context::TyCtxt>::par_hir_body_owners::<rustc_hir_analysis[4c40a9e8caa59823]::check_crate::{closure#2}>::{closure#0}
  28:     0x7fe74de84993 - rustc_hir_analysis[4c40a9e8caa59823]::check_crate
  29:     0x7fe74d506413 - rustc_interface[593877ef20606c3d]::passes::analysis
  30:     0x7fe74e671309 - rustc_query_impl[c3886bc306851868]::execution::try_execute_query::<rustc_middle[1620aa76878d0fde]::query::caches::SingleCache<rustc_middle[1620aa76878d0fde]::query::erase::ErasedData<[u8; 0usize]>>, false>
  31:     0x7fe74e670f75 - rustc_query_impl[c3886bc306851868]::query_impl::analysis::execute_query_non_incr::__rust_end_short_backtrace
  32:     0x7fe74e619a08 - rustc_interface[593877ef20606c3d]::interface::run_compiler::<(), rustc_driver_impl[12052c784fa38642]::run_compiler::{closure#0}>::{closure#1}
  33:     0x7fe74e64533e - std[f332ca7ac2aab280]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[593877ef20606c3d]::util::run_in_thread_with_globals<rustc_interface[593877ef20606c3d]::util::run_in_thread_pool_with_globals<rustc_interface[593877ef20606c3d]::interface::run_compiler<(), rustc_driver_impl[12052c784fa38642]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  34:     0x7fe74e645be0 - <std[f332ca7ac2aab280]::thread::lifecycle::spawn_unchecked<rustc_interface[593877ef20606c3d]::util::run_in_thread_with_globals<rustc_interface[593877ef20606c3d]::util::run_in_thread_pool_with_globals<rustc_interface[593877ef20606c3d]::interface::run_compiler<(), rustc_driver_impl[12052c784fa38642]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[caaf70a19b46f696]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  35:     0x7fe74e646a6c - <std[f332ca7ac2aab280]::sys::thread::unix::Thread>::new::thread_start
  36:     0x7fe7480a598b - <unknown>
  37:     0x7fe748129a0c - <unknown>
  38:                0x0 - <unknown>

error: the compiler unexpectedly panicked. This is a bug

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.96.0-nightly (a0e206b56 2026-03-13) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `main`
#1 [analysis] running analysis passes on crate `mvce`
end of query stack
error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0658`.

```

</p>
</details>

<!--
query stack:
#0 [typeck] type-checking `main`
#1 [analysis] running analysis passes on crate `mvce`

-->
@rustbot label +F-unboxed_closures

--- a/compiler/rustc_middle/src/ty/print/pretty.rs
+++ b/compiler/rustc_middle/src/ty/print/pretty.rs
@@ -3314,7 +3314,8 @@ define_print_and_forward_display! {
     TraitRefPrintSugared<'tcx> {
         if !with_reduced_queries()
             && p.tcx().trait_def(self.0.def_id).paren_sugar
-            && let ty::Tuple(args) = self.0.args.type_at(1).kind()
+            && let Some(args_ty) = self.0.args.get(1).and_then(|arg| arg.as_type())
+            && let ty::Tuple(args) = args_ty.kind()
         {
             write!(p, "{}(", p.tcx().item_name(self.0.def_id))?;
             for (i, arg) in args.iter().enumerate() {

--- a/tests/ui/unboxed-closures/paren-sugar-non-fn-trait-issue-153855.rs
+++ b/tests/ui/unboxed-closures/paren-sugar-non-fn-trait-issue-153855.rs
@@ -1,0 +1,12 @@
+#![feature(rustc_attrs, unboxed_closures)]
+#![allow(internal_features)]
+
+#[rustc_paren_sugar]
+trait Tr {
+    fn method();
+}
+
+fn main() {
+    <u8 as Tr>::method();
+    //~^ ERROR the trait bound `u8: Tr` is not satisfied
+}

--- a/tests/ui/unboxed-closures/paren-sugar-non-fn-trait-issue-153855.stderr
+++ b/tests/ui/unboxed-closures/paren-sugar-non-fn-trait-issue-153855.stderr
@@ -1,0 +1,15 @@
+error[E0277]: the trait bound `u8: Tr` is not satisfied
+  --> $DIR/paren-sugar-non-fn-trait-issue-153855.rs:10:6
+   |
+LL |     <u8 as Tr>::method();
+   |      ^^ the trait `Tr` is not implemented for `u8`
+   |
+help: this trait has no implementations, consider adding one
+  --> $DIR/paren-sugar-non-fn-trait-issue-153855.rs:5:1
+   |
+LL | trait Tr {
+   | ^^^^^^^^
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0277`.

--- a/tests/ui/unboxed-closures/paren-sugar-non-fn-trait-ufcs-issue-153855.rs
+++ b/tests/ui/unboxed-closures/paren-sugar-non-fn-trait-ufcs-issue-153855.rs
@@ -1,0 +1,16 @@
+#![feature(rustc_attrs, unboxed_closures)]
+#![allow(internal_features)]
+
+// Regression test for #153855: diagnostics should not ICE when a
+// `#[rustc_paren_sugar]` trait is used with parenthesized syntax but does not
+// have an `Fn`-family generic layout.
+#[rustc_paren_sugar]
+trait Tr<'a, 'b, T> {
+    fn method() {}
+}
+
+fn main() {
+    <u8 as Tr(&u8)>::method;
+    //~^ ERROR associated item constraints are not allowed here
+    //~| ERROR the trait bound `u8: Tr<'_, '_, (&u8,)>` is not satisfied
+}

--- a/tests/ui/unboxed-closures/paren-sugar-non-fn-trait-ufcs-issue-153855.stderr
+++ b/tests/ui/unboxed-closures/paren-sugar-non-fn-trait-ufcs-issue-153855.stderr
@@ -1,0 +1,22 @@
+error[E0229]: associated item constraints are not allowed here
+  --> $DIR/paren-sugar-non-fn-trait-ufcs-issue-153855.rs:13:12
+   |
+LL |     <u8 as Tr(&u8)>::method;
+   |            ^^^^^^^ associated item constraint not allowed here
+
+error[E0277]: the trait bound `u8: Tr<'_, '_, (&u8,)>` is not satisfied
+  --> $DIR/paren-sugar-non-fn-trait-ufcs-issue-153855.rs:13:6
+   |
+LL |     <u8 as Tr(&u8)>::method;
+   |      ^^ the trait `Tr<'_, '_, (&u8,)>` is not implemented for `u8`
+   |
+help: this trait has no implementations, consider adding one
+  --> $DIR/paren-sugar-non-fn-trait-ufcs-issue-153855.rs:8:1
+   |
+LL | trait Tr<'a, 'b, T> {
+   | ^^^^^^^^^^^^^^^^^^^
+
+error: aborting due to 2 previous errors
+
+Some errors have detailed explanations: E0229, E0277.
+For more information about an error, try `rustc --explain E0229`.

